mimxrt: Rework flash configuration.

alphaFred opened the pull request:
- Moves definition of BOARD_FLASH_SIZE and other header files related to
	flash configuration into the Makefile.
- Adds board specific clock_config.h.
- Adds board.h, pin_mux.h, and peripherals.h as they are
	required by NXP MCU SDK in order to use our own clock_config.h.
- Renames board specific FlexSPI configuration files.
- Updates flash frequency of MIMXRT1020_EVK
- Creates separated flash_config files for QSPI NOR and
	QSPI Hyper flash.
- Unifies VFS start address to be @ 1M for 1010 and 1020 boards.
- Unifies 1050EVK boards
- Adds support to both NOR and HyperFlash on boards with
	both capabilities.
- Adds automatic FlexRAM initialization to start-up code based on
	linker script and NXP HAL.

With this change the flash configuration is restructured. 
This simplifies the configuration process and
provides a better overview of each board's settings. With the integration
of clock_config.h, board.h, pin_mux.h, and peripherals.h we gain better
control of the settings and clock configurations. Furthermore the
implementation of an explicit FlexRAM setup improves the system
performance and allows for performance tuning.

Signed-off-by: Philipp Ebensberger

alphaFred committed:
mimxrt: Rework flash configuration.

- Moves definition of BOARD_FLASH_SIZE and other header files related to
	flash configuration into the Makefile.
- Adds board specific clock_config.h.
- Adds board.h, pin_mux.h, and peripherals.h as they are
	required by NXP MCU SDK in order to use our own clock_config.h.
- Renames board specific FlexSPI configuration files.
- Updates flash frequency of MIMXRT1020_EVK
- Creates separated flash_config files for QSPI NOR and
	QSPI Hyper flash.
- Unifies VFS start address to be @ 1M for 1010 and 1020 boards.
- Unifies 1050EVK boards
- Adds support to both NOR and HyperFlash on boards with
	both capabilities.
- Adds automatic FlexRAM initialization to start-up code based on
	linker script and NXP HAL.
- Applies code formatting to all files in mimxrt port.

With this change the flash configuration is restructured and
organized. This simplifies the configuration process and
provides a better overview of each board's settings. With the integration
of clock_config.h, board.h, pin_mux.h, and peripherals.h we gain better
control of the settings and clock configurations. Furthermore the
implementation of an explicit FlexRAM setup improves the system
performance and allows for performance tuning.

Signed-off-by: Philipp Ebensberger

robert-hh commented:
Just a few additional notes on that PR:
This PR looks huge, but is an attempt to clean up the various board file trees to a common structure. Over the time, things got a little bit fuzzy. And so we decided to make this PR. It covers the fact, that the boards can be configured with different flash setting. It also includes a change to make better use of the built-in RAM. That causes most of the added lines of code, because it moves the startup_xxx.s files from the nxp_lib to the port file system. That seems necessary since the RAM re-config has to happen before any RAM is used.

dpgeorge commented on ports/mimxrt/boards/make-flexram-config.py:
```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python
+"""Evaluate FlexRAM configuration and generate startup code."""
```
what is the license/copyright of this file?

robert-hh replied:
It was written by @alphaFred, so I assume it's MIT, and Philipp will add the appropriate License.

alphaFred replied:
Yes - I will add the license

dpgeorge replied:
please make the style of the license/copyright match existing files, eg `tools/pyboard.py` and `tools/mpy_ld.py` (use `#` at the start of the line)

dpgeorge commented:
Regarding the startup_MIMXRTxxx.S files: in the HAL versions of these files the Reset_Handler function is defined as a weak symbol.  So that means you can reimplement just this function in your own code, which is exactly what you need.  This is how stm32 works, it uses the HAL's startup code to get the IRQ table, but reimplements Reset_Handler in `ports/stm32/resethandler.s`.  I'd suggest mimxrt also do it this way so that there is less code, and maybe the same resethandler.s (or .S if it needs to use the C preprocessor) can be used for all MCU variants.

Also, why exactly does the `clock_config.h` file need to be copied out of the HAL into each board directory?  As far as I can tell it's copied without change, so why can't it reuse the HAL version?

alphaFred commented:
Hi @dpgeorge. Thanks for your review. 

Regarding your first point - I will give it a try and check if it works for us.

Regarding `clock_config.h`. This is part of a set of four files (`clock_config.h`, `board.h`,`peripherals.h`,`pin_mux.h`) which are provided as templates. So far this worked fine for us but with this cleanup I noticed that the `board.h` defines `BOARD_FLASH_SIZE` with out `#ifdef`guard. This is a problem when we want to configure for different flash sizes (e.g. QPSI vs. HyperFlash). To get around that I needed to create my own set of these header files. Since they are all located in one folder I could not easily exclude `clock_config.h`. 

On a side note: 
I already have another PR ready which builds on top of this one where we will refactor the clock configuration. in that PR you will see differences in `clock_config.h`.

robert-hh commented:
> maybe the same resethandler.s (or .S if it needs to use the C preprocessor) can be used for all MCU variants.

That is indeed the case. I just tested that (and compared the files) and it works for all MCU's I have here.

dpgeorge commented on ports/mimxrt/hal/board.h:
```diff
@@ -0,0 +1,16 @@
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
```
This file is pretty much written from scratch, so I think you should use your own license/copyright.  And it's not a dummy file because it includes things that are actually needed!

dpgeorge commented on ports/mimxrt/hal/board.h:
```diff
@@ -0,0 +1,16 @@
+ */
+
+#ifndef _BOARD_H_
+#define _BOARD_H_
```
Please match other include guard names, `MICROPY_INCLUDED_MIMXRT_BOARD_H`

dpgeorge commented on ports/mimxrt/hal/peripherals.h:
```diff
@@ -0,0 +1,8 @@
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
```
This is an empty file so doesn't need a license/copyright.  It can just be:
```c
// Empty file, necessary for compilation with NXP MCU SDK
```

dpgeorge commented on ports/mimxrt/hal/pin_mux.h:
```diff
@@ -0,0 +1,8 @@
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
```
same as above

dpgeorge commented on ports/mimxrt/hal/startup_MIMXRT10xx.S:
```diff
@@ -0,0 +1,166 @@
+
```
I'd call this file `resethandler_MIMXRT10xx.S`.  Also it'll need copyright/license copied from the file it was taken from, because there's a lot of copied code.

dpgeorge commented on ports/mimxrt/hal/startup_MIMXRT10xx.S:
```diff
@@ -0,0 +1,166 @@
+    ldr    r2, =__data_start__
+    ldr    r3, =__data_end__
+
+#ifdef __PERFORMANCE_IMPLEMENTATION
```
The difference in code size between `__PERFORMANCE_IMPLEMENTATION` enabled and disabled is very small, and these chips have a huge amount of flash, so I think it's simpler/cleaner to just unconditionally use the performance version of the code (ie remove the other code).

alphaFred replied:
Hi, I tried it out now but it does not work. Even when just compiling with `__PERFORMANCE_IMPLEMENTATION` defined, MicroPython crashes.
If you are ok with it, let us merge it as it is and we will look into it later and fix it.

alphaFred commented:
@dpgeorge  I pushed an update with the required fixes (:smile: )

alphaFred commented:
@dpgeorge @robert-hh 
I pushed an update with an update to code formatting (added `"ports/mimxrt/**/*.[ch]",` to `tools/codeformat.py`as in #7785) and ran the formatting on all the files in mimxrt port. Should be indentical with #7785 now.

Btw. whenever I execute `codeformat.py` it also creates changes in other files:
```
	modified:   ../../extmod/machine_i2c.c
	modified:   ../../extmod/vfs_lfsx.c
	modified:   ../stm32/mpconfigport.h
	modified:   ../unix/mpconfigport.h
	modified:   ../../py/compile.
```

dpgeorge commented:
> Btw. whenever I execute `codeformat.py` it also creates changes in other files:

It's not showing on the CI here, so could be your local uncrustify version.

alphaFred commented:
Ok, then never mind :-)

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -50,12 +63,12 @@ INC += -I$(BUILD)
 INC += -I$(TOP)/lib/tinyusb/hw
 INC += -I$(TOP)/lib/tinyusb/hw/bsp/teensy_40
 INC += -I$(TOP)/lib/tinyusb/src
+INC += -Ihal
```
please move this up to just below `-I.`, so that current-dir includes are before other ones

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -277,9 +312,10 @@ ifeq ($(MICROPY_FLOAT_IMPL),double)
+SRC_SS = \
+	$(MCU_DIR)/gcc/startup_$(MCU_SERIES).S \
+	hal/resethandler_MIMXRT10xx.S
+SRC_S = shared/runtime/gchelper_m3.s \
```
this `SRC_S` line didn't need to change (should still be +=)

also please retain the blank line between them

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -316,8 +352,10 @@ $(BUILD)/lib/tinyusb/src/device/usbd.o: CFLAGS += -Wno-missing-braces
+	$(Q)$(CC) -E -x c $(LDDEFINES) $(LD_FILES) | grep -v '^#' > $(BUILD)/link.ld
 	$(ECHO) "LINK $@"
-	$(Q)$(LD) $(LDFLAGS) -o $@ $^ $(LIBS)
+	$(Q)$(LD) $(addprefix -T, $(BUILD)/link.ld)  $(LDFLAGS) -o $@ $^ $(LIBS)
```
can just use `-T$(BUILD)/link.ld`, no addprefix needed

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -326,17 +364,23 @@ $(BUILD)/firmware.bin: $(BUILD)/firmware.elf
 
+$(GEN_FLEXRAM_CONFIG_SRC):
+	$(ECHO) "Create $@"
+	$(Q)$(PYTHON) $(MAKE_FLEXRAM_LD) -d $(TOP)/$(MCU_DIR)/$(MCU_SERIES).h\
```
please put a space before the `\` at the end

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1050_EVK/mpconfigboard.h:
```diff
@@ -1,14 +1,10 @@
-#define MICROPY_HW_BOARD_NAME "i.MX RT1050 EVK"
+#define MICROPY_HW_BOARD_NAME "i.MX RT1050 EVKB"
```
It looks like the board dir got renamed from MIMXRT1050_EVKB to MIMXRT1050_EVK, yet this name here is the opposite, the B got added, why?

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1052.ld:
```diff
@@ -1,14 +1,25 @@
+#else
+#error Unknown BOARD_FLASH_TYPE
+#endif
+flash_size 			= BOARD_FLASH_SIZE;
```
looks like there are tabs in this line, please replace them with spaces

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1062.ld:
```diff
@@ -1,14 +1,25 @@
+#else
+#error Unknown BOARD_FLASH_TYPE
+#endif
+flash_size 			= BOARD_FLASH_SIZE;
```
please replace tab with spaces

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1064.ld:
```diff
@@ -1,14 +1,27 @@
+#else
+#error Unknown BOARD_FLASH_TYPE
+#endif
+flash_size 			= BOARD_FLASH_SIZE;
```
tabs to spaces

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/hal/flexspi_hyper_flash.c:
```diff
@@ -44,17 +45,17 @@ __attribute__((always_inline)) static inline void clock_disable_clock(clock_ip_n
-#define DIV_READ		0
+#define DIV_PAGE_PGM    4
+#define DIV_ERASE_PGM   4
+#define DIV_READ                0
```
please align these values to the same column

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1021.ld:
```diff
@@ -1,20 +1,29 @@
+#else
+#error Unknown BOARD_FLASH_TYPE
+#endif
+flash_size 			= BOARD_FLASH_SIZE;
```
tabs to spaces

alphaFred replied:
fixed.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1011.ld:
```diff
@@ -1,15 +1,24 @@
+#else
+#error Unknown BOARD_FLASH_TYPE
+#endif
+flash_size 			= BOARD_FLASH_SIZE;
```
tabs to spaces

alphaFred replied:
fixed.

dpgeorge commented:
@alphaFred at this stage I'd still need to rebase it and reword the commit message (to something like `mimxrt: Rework flash configuration." to get the tense correct).  But feel free to do that yourself if you like.

alphaFred commented:
Ok. I hope I get the tense right the next time at first try 😂.

alphaFred commented:
@dpgeorge 
Ok, I fixed your latest comments, rebased and fixed the title of the commit. Should be good to go. Many thanks for your quick review.

dpgeorge commented:
Merged!
